packages/contracts/src/Game.sol

```diff
@@ -467,9 +467,59 @@ contract Game {
     // (swap removed card with last card, and truncate the array by one).
     //
     // The player's deck is cards[pdata.deckStart:pdata.deckEnd].
-    function checkInitialHandProof(PlayerData storage pdata, uint256[] storage cards,
+    function checkInitialHandProof(PlayerData storage pdata, bytes32 initialDeckRoot,
              uint256 randomness, bytes calldata proof) view internal {
-        // TODO(PROOF)
+        // use randomness to decide drawn cards
+        uint256[7] memory selectedIndices = drawRandomIndices(randomness);
+
+        // construct circuit public signals
+        uint256[] memory pubSignals = new uint256[](131);
+        pubSignals[0] = uint256(initialDeckRoot);
+        pubSignals[1] = uint256(pdata.deckRoot);
+        pubSignals[2] = uint256(pdata.handRoot);
+        // elements 67-130 are the binary drawn indices (0 if not drawn, 1 if drawn)
+        for (uint256 i = 0; i < 7; i++) {
+            pubSignals[selectedIndices[i] + 67] = 1;
+        }
+        // elements 3-66 are the deck predicates
+        uint256 deckCount = 0;
+        uint256 handCount = 0;
+        for (uint256 i = 3; i < 67; i++) {
+            if (pubSignals[i+64] == 1) {
+                // the index is drawn
+                pubSignals[i] = 2**handCount;
+                handCount++;
+            } else {
+                // the index is not drawn
+                pubSignals[i] = 2**deckCount;
+                deckCount++;
+            }
+        }
+        initialVerifier.verifyProof(proof, pubSignals);
+    }
+
+    function drawRandomIndices(uint256 randomness) internal pure returns (uint256[7] memory) {
+        uint256[64] memory indices;
+        uint256[7] memory selectedIndices;
+        
+        // Initialize the elements array with values 0-63
+        for (uint256 i = 0; i < 64; i++) {
+            indices[i] = i;
+        }
+        
+        // Draw 7 random elements
+        uint256 lastIndex = 63;
+        for (uint256 i = 0; i < 7; i++) {
+            // Generate a random index within the remaining elements
+            uint256 randomIndex = uint256(keccak256(abi.encodePacked(randomness, i))) % (lastIndex + 1);
+            
+            // Select the random element and move the last element to its place
+            selectedIndices[i] = indices[randomIndex];
+            indices[randomIndex] = indices[lastIndex];
+            lastIndex--;
+        }
+        
+        return selectedIndices;
```

Comment:
This cannot actually live in the contract and must be implemented in the circuit side. The reason is that knowledge of the drawn indices means the opponent would know the deck.

To make this secure, we actually need the user to also pick a salt, and submit the hash of the salt in a first tx. Then later he can submit the proof, where the randomness (computed on the circuit side) will be based on both the salt that has been committed to and the randomness picked from the contract by using the hash for the last block number.

(It's crucial that the user does not know which hash will be used when he picks his salt, otherwise he could grind the salt to get the hand he wants.)

The circuit will take the salt as private input, its hash as public input and verify adequation. It also take "the public randomness" as public input, and computes the cards to draw based on this.

Reply:
I see, my initial thought is that when a player has its deck at the start of the game, it will first shuffle the deck to hide the ordering from the opponent (there's a `shuffle` circuit to make sure the player has shuffle correctly, the player could shuffle arbitrarily). Then, we will draw the cards randomly and publicly on the smart contract. 

I think what you mentioned also works, but implementing the fisher yates algorithm to select cards in the circuit could be expensive, since it involves accessing an unknown index during compilation. This usually requires looping through the entire array for each selection. So I was thinking it might be easier to shuffle off-chain, verify it through the `shuffle` circuit and draw the indices publicly. But I guess this is dependent on the desired game flow, so could also make the changes. Let me know your thoughts!

Reply:
Oh lord, I don't remember the `Shuffle` circuit at all :')

You don't need to shuffle at all though! If your indices are private (which they are if you use the salt in the randomness) then it doesn't matter that the deck is not initially shuffled — there is not way for your opponent to know what you drew.

(Now that I think of it, he might be able to brute force the Merkle root, so it should probably be hashed with the salt before being posted on chain too.)

Reply:
But to prevent for selecting the same indices twice, we still need to use the fisher yates algorithm to move the selected card to the last index? Essentially I think this [function](https://github.com/norswap/0xFable/blob/initial-hand-verification/packages/contracts/src/Game.sol#L501) is a form of fisher yates, and if we move this function to the circuit, we'll need to: 
- introduce `signal` of size `7 x 64`, where each row is used for each card selection
- for each selection, loop through the entire deck of 64 cards

Reply:
Yes, that seems to be correct. There might be a better solution, but it there is I can't think of it right now!

There might be hope if we can do the "bitfield trick" where we represent an array of size `S <= 256` of items of magnitude `M < 256` as a bitfield and can use that bitfield to test if an item is in the array in a constant number of constraints (vs iterating over the whole array, which we would have had to do with no bitfield). Did we ever try if that works?

(Still, even if that's possible, the solution is still distinctively non-trivial.)